File: strings/271_encode_and_decode_strings.py
# Approach 1: Encode and Decode
# Time complexity: O(m) for each encode() and decode(). m is sum of lengths of all strings.
# Space complexity: O(n + m) for encode() and decode(). n is number of strings

class Solution:

    def encode(self, strs: List[str]) -> str:
        encoded = ""
        for s in strs:
            encoded = encoded + str(len(s)) + "#" + s
        return encoded
        
    def decode(self, s: str) -> List[str]:
        res = []
        i = 0

        while i < len(s):
            j = i
            while s[j] != "#":
                j += 1
            length = int(s[i:j])
            i = j + 1
            j = i + length
            res.append(s[i:j])
            i = j
        return res

File: strings/271_encode_and_decode_strings/271_encode_and_decode_strings.java
// Approach 1: Encode and Decode
// Time complexity: O(m) for each encode() and decode(). m is sum of lengths of all strings.
// Space complexity: O(n + m) for encode() and decode(). n is number of strings.

class Solution {

    public String encode(List<String> strs) {
        StringBuilder res = new StringBuilder();
        for (String s : strs) {
            res.append(s.length()).append('#').append(s);
        }
        return res.toString();
    }

    public List<String> decode(String str) {
        List<String> res = new ArrayList<>();
        int i = 0;

        while(i < str.length()) {
            int j = i;
            while (str.charAt(j) != '#') {
                j++;
            }
            int length = Integer.parseInt((str.substring(i, j)));
            i = j + 1;
            j = i + length;
            res.add(str.substring(i, j));
            i = j;
        }
        return res;
    }
}
